docs/config_options.md
# Klayr Service Configuration Reference

## Common settings

These options are applicable to all the Klayr Service microservices.
When using the Docker setup, the default values specified in the [`docker/example.env`](../docker/example.env) file should suffice. To properly set the Docker environment variables, copy the [`docker/example.env`](../docker/example.env) file as `.env` and place it next to the [`docker-compose.yml`](../docker-compose.yml) file in the project's root directory. Please adjust the values for the environment variables, if necessary.

> **Note:**  If the Docker environment is being used, the minimum necessary environment variable that needs to be set is `KLAYR_APP_URLS`.

### Service broker

```bash
# Must be identical for all microservices
# Make sure that all microservices are able to connect with the common Redis
SERVICE_BROKER=redis://klayr:password@127.0.0.1:6379/0

# Number of seconds to wait before returning a RequestTimeout error when it takes too long to return a value. To disable use 0.
SERVICE_BROKER_TIMEOUT=10
```
### Log configuration

```bash
SERVICE_LOG_STDOUT=true   # Asynchronous console output (non-blocking, preferred)
SERVICE_LOG_CONSOLE=false # console.log() output, only for debug
SERVICE_LOG_FILE=false    # file path ex. ./logs/service.log
SERVICE_LOG_GELF=false    # GELF output for remote logging ex. Graylog 127.0.0.1:12201/udp
SERVICE_LOG_LEVEL=info    # Default log level
DOCKER_HOST=local         # Custom field for logger. This will result in all log messages having the custom field _docker_host set to 'local'.
```

## Microservice-specific settings

Every Klayr Service microservice accepts various environment variables, that can be set to modify the application's behavior.\
For an exhaustive list of the supported environment variables, please check the `Configuration` section in the microservice specific README files.

The following lists all the available microservices and links to their documentation. Please follow the links to learn more about the available configurable environment variables.

1. [Gateway](../services/gateway/README.md#configuration)
2. [Connector](../services/blockchain-connector/README.md#configuration)
3. [Coordinator](../services/blockchain-coordinator/README.md#configuration)
4. [Indexer](../services/blockchain-indexer/README.md#configuration)
5. [App Registry](../services/blockchain-app-registry/README.md#configuration)
6. [Fee Estimator](../services/fee-estimator/README.md#configuration)
7. [Transaction Statistics](../services/transaction-statistics/README.md#configuration)
8. [Market](../services/market/README.md#configuration)
9. [Export](../services/export/README.md#configuration)


services/market/config.js
/*
 * Klayrhq/klayrservice
 * Copyright © 2021 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */
const packageJson = require('./package.json');

const config = {
	endpoints: {},
	market: {},
};

// Moleculer broker config
config.transporter = process.env.SERVICE_BROKER || 'redis://klayr:password@127.0.0.1:6379/0';
config.brokerTimeout = Number(process.env.SERVICE_BROKER_TIMEOUT) || 10; // in seconds

// Logging
config.log = {
	name: packageJson.name,
	version: packageJson.version,
};
/**
 * log.level - Limits the importance of log messages for console and stdout outputs
 *             One fo the following in that order:
 *               TRACE < DEBUG < INFO < WARN < ERROR < FATAL < MARK
 */
config.log.level = process.env.SERVICE_LOG_LEVEL || 'info';

/*
 * True / False outputs
 * log.console - Plain JavaScript console.log() output
 * log.stdout  - Writes directly to stdout
 */
config.log.console = process.env.SERVICE_LOG_CONSOLE || 'false';
config.log.stdout = process.env.SERVICE_LOG_STDOUT || 'true';

/*
 * Configurable outputs
 * log.file   - outputs to a file (ie. ./logs/app.log)
 * log.gelf   - Writes to GELF-compatible socket (ie. 127.0.0.1:12201/udp)
 */
config.log.gelf = process.env.SERVICE_LOG_GELF || 'false';
config.log.file = process.env.SERVICE_LOG_FILE || 'false';

// Set docker host if running inside the container
config.log.docker_host = process.env.DOCKER_HOST || 'local';

// Api keys to access apis
config.access_key = {};
config.access_key.exchangeratesapi = process.env.EXCHANGERATESAPI_IO_API_KEY;

// Expiry time for redis
config.ttl = {
	exchangeratesapi: 24 * 60 * 60 * 1000, // milliseconds,
	binance: 15 * 60 * 1000, // milliseconds
	bitrue: 15 * 60 * 1000, // milliseconds
	probit: 15 * 60 * 1000, // milliseconds
};

/**
 * External endpoints
 */
config.endpoints.redis =
	process.env.SERVICE_MARKET_REDIS || 'redis://klayr:password@127.0.0.1:6379/6';

/**
 * Market prices config
 */
// SERVICE_MARKET_FIAT_CURRENCIES & SERVICE_MARKET_TARGET_PAIRS should be CSV-based strings
config.market.supportedFiatCurrencies =
	process.env.SERVICE_MARKET_FIAT_CURRENCIES || 'EUR,USD,CHF,GBP,RUB,PLN,JPY,AUD,GBP,INR';
config.market.targetPairs =
	process.env.SERVICE_MARKET_TARGET_PAIRS ||
	'KLY_BTC,KLY_EUR,KLY_USD,KLY_CHF,KLY_PLN,KLY_JPY,KLY_AUD,KLY_GBP,KLY_INR,BTC_EUR,BTC_USD,BTC_CHF';
config.market.sources = {
	bitrue: {
		allowRefreshAfter: 1 * 60 * 1000, // milliseconds
	},
	probit: {
		allowRefreshAfter: 1 * 60 * 1000, // milliseconds
	},
	binance: {
		apiEndpoint: 'https://api.binance.com/api/v3',
		allowRefreshAfter: 1 * 60 * 1000, // milliseconds
	},
	exchangeratesapi: {
		apiEndpoint: 'http://api.exchangeratesapi.io/v1',
		allowRefreshAfter: 8 * 60 * 60 * 1000, // milliseconds
	},
};

config.job = {
	// Interval takes priority over schedule and must be greater than 0 to be valid
	refreshPricesBitrue: {
		interval: Number(process.env.JOB_INTERVAL_REFRESH_PRICES_BITRUE) || 0,
		schedule: process.env.JOB_SCHEDULE_REFRESH_PRICES_BITRUE || '* * * * *',
	},
	refreshPricesProbit: {
		interval: Number(process.env.JOB_INTERVAL_REFRESH_PRICES_PROBIT) || 0,
		schedule: process.env.JOB_SCHEDULE_REFRESH_PRICES_PROBIT || '* * * * *',
	},
	refreshPricesBinance: {
		interval: Number(process.env.JOB_INTERVAL_REFRESH_PRICES_BINANCE) || 0,
		schedule: process.env.JOB_SCHEDULE_REFRESH_PRICES_BINANCE || '* * * * *',
	},
	refreshPricesExchangeratesapi: {
		interval: Number(process.env.JOB_INTERVAL_REFRESH_PRICES_EXCHANGERATESAPI) || 0,
		schedule: process.env.JOB_SCHEDULE_REFRESH_PRICES_EXCHANGERATESAPI || '* * * * *',
	},
	updatePrices: {
		interval: Number(process.env.JOB_INTERVAL_UPDATE_PRICES) || 5,
		schedule: process.env.JOB_SCHEDULE_UPDATE_PRICES || '',
	},
};

module.exports = config;
